internal/fields/validate.go

```diff
@@ -0,0 +1,217 @@
+		err = yaml.Unmarshal(body, &u)
+		if err != nil {
+			return nil, errors.Wrap(err, "unmarshalling field body failed")
+		}
+		fields = append(fields, u...)
+	}
+	return &Validator{
+		schema: fields,
+	}, nil
+}
+
+// ValidateDocumentBody validates the provided document body.
+func (v *Validator) ValidateDocumentBody(body json.RawMessage) multierror.Error {
+	var c common.MapStr
+	err := json.Unmarshal(body, &c)
+	if err != nil {
+		var errs multierror.Error
+		errs = append(errs, errors.Wrap(err, "unmarshalling document body failed"))
+		return errs
+	}
+
+	errs := v.validateMapElement("", c)
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// ValidateDocumentMap validates the provided document as common.MapStr.
+func (v *Validator) ValidateDocumentMap(body common.MapStr) multierror.Error {
+	errs := v.validateMapElement("", body)
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+func (v *Validator) validateMapElement(root string, elem common.MapStr) multierror.Error {
+	var errs multierror.Error
+	for name, val := range elem {
+		key := strings.TrimLeft(root+"."+name, ".")
+
+		switch val.(type) {
+		case []map[string]interface{}:
+			for _, m := range val.([]map[string]interface{}) {
+				err := v.validateMapElement(key, m)
+				if err != nil {
+					errs = append(errs, err...)
+				}
+			}
+		case map[string]interface{}:
+			err := v.validateMapElement(key, val.(map[string]interface{}))
+			if err != nil {
+				errs = append(errs, err...)
+			}
+		default:
+			err := v.validateScalarElement(key, val)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return errs
+}
+
+func (v *Validator) validateScalarElement(key string, val interface{}) error {
+	if key == "" {
+		return nil // root key is always valid
+	}
+
+	definition := findElementDefinition("", key, v.schema)
+	if definition == nil && skipValidationForField(key) {
+		return nil // generic field, let's skip validation for now
+	}
+	if definition == nil {
+		return fmt.Errorf(`field "%s" is undefined`, key)
+	}
+
+	err := parseElementValue(key, *definition, val)
+	if err != nil {
+		return errors.Wrap(err, "parsing field value failed")
+	}
+	return nil
+}
+
+// skipValidationForField skips field validation (field presence) of special fields. The special fields are present
+// in every (most?) documents collected by Elastic Agent, but aren't defined in any integration in `fields.yml` files.
+// FIXME https://github.com/elastic/elastic-package/issues/147
+func skipValidationForField(key string) bool {
+	return isFieldFamilyMatching("agent", key) ||
+		isFieldFamilyMatching("elastic_agent", key) ||
+		isFieldFamilyMatching("cloud", key) || // too many common fields
+		isFieldFamilyMatching("event", key) || // too many common fields
+		isFieldFamilyMatching("host", key) || // too many common fields
+		isFieldFamilyMatching("metricset", key) || // field is deprecated
+		isFieldFamilyMatching("event.module", key) // field is deprecated
```

Comment:
This line is now redundant with `isFieldFamilyMatching("event", key)` but I think we should keep it anyway because even if we were to somehow get rid of `isFieldFamilyMatching("event", key)` in the future, we would want this line to be here, correct?

Reply:
Right, maybe it's better to keep it here even though it's redundant.